### PR TITLE
[dynamo] Enable typechecking for utils.py

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112999
* #112975
* #112974
* #112973
* #112972
* __->__ #112971
* #112970
* #112130

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
Minor comments, otherwise great stuff.

It'd be good not to change too mucht the NumPy import pattern, as we pretty much use if all throughout PyTorch.

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -72,17 +95,17 @@
 from torch.utils._pytree import tree_map_only
 
 
-counters = collections.defaultdict(collections.Counter)
+counters: DefaultDict[str, Counter[str]] = collections.defaultdict(collections.Counter)
```
[the remark this answers is not in the document]
Comment (int3) [reply torch/_dynamo/utils.py:98]: it can't infer the generic types from this definition

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -28,12 +28,30 @@
...
+    Union,
+    ValuesView,
+)
 
+np: Optional[types.ModuleType] = None
 try:
     import numpy as np
 except ModuleNotFoundError:
-    np = None
+    pass
```
[the remark this answers is not in the document]
Comment (int3) [reply torch/_dynamo/utils.py:54]: type tag doesn't work -- we can't specify the type for `np` in the import statement, and we can't type tag the exception handler definition because mypy considers it already defined in the try block. I guess we can use a `type: ignore` comment though...

File: torch/_dynamo/utils.py [hunk 3]
```diff
@@ -44,7 +62,12 @@
 
     # NOTE: Make sure `NP_SUPPORTED_MODULES` and `NP_TO_TNP_MODULE` are in sync.
     if np:
-        NP_SUPPORTED_MODULES = (np, np.fft, np.linalg, np.random)
+        NP_SUPPORTED_MODULES: Tuple[types.ModuleType, ...] = (
```
[the remark this answers is not in the document]
Comment (int3) [reply torch/_dynamo/utils.py:65]: The inferred type here is `Tuple[Module, Any, Any, Any]`, which conflicts with the inferred type in the `else` block which is `Tuple[]`. `NP_TO_TNP_MODULE` doesn't have the same issue as mypy doesn't encode the number of elements of the dict in its type.

File: torch/_dynamo/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (int3) [reply torch/_dynamo/utils.py:54]: One advantage of the type-ignore approach is that mypy will simply treat `np` as a non-optional, successful import, and look up the types accordingly for attributes within that module. Explicitly typing `np` as `Optional[types.ModuleType]` will make mypy treat `np` as a generic module that can have any attributes.

**Review by jansel — APPROVED** [review]
